lib/runner/extract-runnable-items.js

```diff
@@ -0,0 +1,172 @@
+             * @param {String} id
+             *
+             * @returns {Boolean|undefined}
+             */
+            idLookup = function (itemGroup, id) {
+                if (!itemGroup || !itemGroup.items) { return; }
+
+                // lookup id on own children first
+                entry = itemGroup.items.one(id);
+
+                // if a match is found stop search and exit
+                if (entry) { return false; }
+
+                // if id does not match any own children start looking up on the grand children
+                itemGroup.items.each(function (itemOrGroup) {
+                    // some of the children maybe items, skip those
+                    if (ItemGroup.isItemGroup(itemOrGroup)) {
+                        return idLookup(itemOrGroup, id);
+                    }
+                });
+            };
+
+        idLookup(collection, id);
+        return entry;
+    },
+
+    /**
+     * Looks for an Item or ItemGroup matching the name on a collection.
+     *
+     * @param {Collection} collection
+     * @param {String} name
+     *
+     * @returns {Item|ItemGroup|undefined}
+     */
+    findItemOrGroupByName = function (collection, name) {
+        var entry,
+
+            /**
+             * Iteratee for looking up an element with given name on an item group.
+             *
+             * @param {ItemGroup} itemGroup
+             * @param {String} name
+             *
+             * @returns {Boolean|undefined}
+             */
+            nameLookup = function (itemGroup, name) {
+                if (!itemGroup || !itemGroup.items) { return; }
+
+                // look for the item or group in own children
+                itemGroup.items.each(function (itemOrGroup) {
+                    if (itemOrGroup.name === name) { entry = itemOrGroup; return false; }
+                });
+
+                // if a match is found on the current level, stop the search and exit
+                if (entry) { return false; }
+
+                // continue search from grand children
+                itemGroup.items.each(function (itemOrGroup) {
+                    // some of the children maybe items
+                    if (ItemGroup.isItemGroup(itemOrGroup)) {
+                        return nameLookup(itemOrGroup, name);
+                    }
+                });
+            };
+
+        nameLookup(collection, name);
+        return entry;
+    },
+
+    /**
+     * Finds an item or group from a path. The path should be an array of ids from the parent chain.
+     *
+     * @param {Collection} collection
+     * @param {Array<String>} path
+     *
+     * @returns {Item|ItemGroup}
+     */
+    findItemOrGroupByPath = function (collection, path) {
+        var entryPath = path || [],
+            entryNode = collection,
+            i,
+            lookupLength = entryPath.length;
+
+        // go down the lookup path
+        for (i = 0; i < lookupLength && entryNode; i++) {
```

Comment:
for (i = 0; (i < lookupLength) && entryNode; i++)

Reply:
also check perf of


```
path = _.clone(path || []);

do {
    collection = collection && collections.items && collection.items.one(path.pop())
} while (collection && path.length);
```

Reply:
Checked the perf difference with `for` and `while with clone and pop`, the `while` options is 60% slower.